File: Backend/app/schemas/market_data.py
import enum
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict


class HistoryPeriod(str, enum.Enum):
    """Supported historical lookback windows. Deliberately a small fixed
    set for this foundation phase -- not arbitrary date-range support."""

    ONE_DAY = "1D"
    ONE_WEEK = "1W"
    ONE_MONTH = "1M"
    THREE_MONTHS = "3M"
    ONE_YEAR = "1Y"


class StockSearchResult(BaseModel):
    """A single search match. Provider-independent: no FYERS-specific
    field names leak out of this schema."""

    model_config = ConfigDict(from_attributes=True)

    symbol: str
    company_name: str
    exchange: str
    # Opaque provider-internal identifier (e.g. the FYERS ticker
    # "NSE:RELIANCE-EQ"). Callers should treat this as a pass-through
    # value, not parse it.
    instrument_key: str


class QuoteResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    symbol: str
    price: Decimal
    previous_close: Decimal | None
    day_high: Decimal | None
    day_low: Decimal | None
    volume: int | None
    average_volume: int | None
    change: Decimal | None
    change_percent: Decimal | None
    timestamp: datetime
    data_source: str
    is_delayed: bool


class HistoricalCandle(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    timestamp: datetime
    open: Decimal
    high: Decimal
    low: Decimal
    close: Decimal
    volume: int


class HistoricalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    symbol: str
    period: HistoryPeriod
    data_source: str
    candles: list[HistoricalCandle]

File: Frontend/src/types/marketData.ts
export type HistoryPeriod = '1D' | '1W' | '1M' | '3M' | '1Y';

export interface StockSearchResult {
  symbol: string;
  company_name: string;
  exchange: string;
  instrument_key: string;
}

/**
 * Decimal-typed backend fields (price, previous_close, day_high, day_low,
 * change, change_percent) arrive as JSON strings, not numbers -- FastAPI
 * serializes Python `Decimal` values as strings to avoid float precision
 * loss. `volume`/`average_volume` are plain ints and stay numbers.
 */
export interface QuoteResponse {
  symbol: string;
  price: string;
  previous_close: string | null;
  day_high: string | null;
  day_low: string | null;
  volume: number | null;
  average_volume: number | null;
  change: string | null;
  change_percent: string | null;
  timestamp: string;
  data_source: string;
  is_delayed: boolean;
}

export interface HistoricalCandle {
  timestamp: string;
  open: string;
  high: string;
  low: string;
  close: string;
  volume: number;
}

export interface HistoricalResponse {
  symbol: string;
  period: HistoryPeriod;
  data_source: string;
  candles: HistoricalCandle[];
}

/** Per-symbol quote-fetch state for the watchlist detail page -- keyed by
 * symbol so one stock's failure never affects another's. */
export type QuoteState =
  | { status: 'loading' }
  | { status: 'success'; data: QuoteResponse }
  | { status: 'error'; message: string };
